File: Store Support/General Setup/Distribution_Lists/api_distribution_lists.js
/**
 * Code Puppy Pages API Endpoint
 * Queries BigQuery for distribution lists
 * 
 * Deploy this to Code Puppy Pages as: /api/distribution-lists
 */

const { BigQuery } = require('@google-cloud/bigquery');

// Initialize BigQuery client
const bigquery = new BigQuery({
    projectId: 'wmt-assetprotection-prod'
});

/**
 * GET /api/distribution-lists
 * Returns all distribution lists from BigQuery
 */
async function getDistributionLists(req, res) {
    try {
        const query = `
            SELECT 
                email,
                name,
                display_name,
                description,
                member_count,
                category
            FROM \`wmt-assetprotection-prod.Store_Support_Dev.dl_catalog\`
            ORDER BY name
        `;
        
        console.log('Executing BigQuery query...');
        const [rows] = await bigquery.query({
            query: query,
            location: 'US'
        });
        
        console.log(`Retrieved ${rows.length} distribution lists`);
        
        res.status(200).json(rows);
        
    } catch (error) {
        console.error('Error querying BigQuery:', error);
        res.status(500).json({
            error: 'Failed to load distribution lists',
            message: error.message
        });
    }
}

/**
 * GET /api/distribution-lists/search?q=searchTerm
 * Search distribution lists by keyword
 */
async function searchDistributionLists(req, res) {
    try {
        const searchTerm = req.query.q || '';
        
        const query = `
            SELECT 
                email,
                name,
                display_name,
                description,
                member_count,
                category
            FROM \`wmt-assetprotection-prod.Store_Support_Dev.dl_catalog\`
            WHERE LOWER(email) LIKE @searchTerm
               OR LOWER(name) LIKE @searchTerm
               OR LOWER(description) LIKE @searchTerm
            ORDER BY name
            LIMIT 100
        `;
        
        const options = {
            query: query,
            location: 'US',
            params: { searchTerm: `%${searchTerm.toLowerCase()}%` }
        };
        
        const [rows] = await bigquery.query(options);
        
        res.status(200).json(rows);
        
    } catch (error) {
        console.error('Error searching BigQuery:', error);
        res.status(500).json({
            error: 'Failed to search distribution lists',
            message: error.message
        });
    }
}

/**
 * GET /api/distribution-lists/category/:category
 * Get distribution lists by category
 */
async function getDistributionListsByCategory(req, res) {
    try {
        const category = req.params.category;
        
        const query = `
            SELECT 
                email,
                name,
                display_name,
                description,
                member_count,
                category
            FROM \`wmt-assetprotection-prod.Store_Support_Dev.dl_catalog\`
            WHERE category = @category
            ORDER BY name
        `;
        
        const options = {
            query: query,
            location: 'US',
            params: { category: category }
        };
        
        const [rows] = await bigquery.query(options);
        
        res.status(200).json(rows);
        
    } catch (error) {
        console.error('Error querying BigQuery:', error);
        res.status(500).json({
            error: 'Failed to load distribution lists',
            message: error.message
        });
    }
}

/**
 * GET /api/distribution-lists/stats
 * Get statistics about distribution lists
 */
async function getDistributionListStats(req, res) {
    try {
        const query = `
            SELECT 
                category,
                COUNT(*) as count,
                SUM(member_count) as total_members,
                AVG(member_count) as avg_members
            FROM \`wmt-assetprotection-prod.Store_Support_Dev.dl_catalog\`
            GROUP BY category
            ORDER BY count DESC
        `;
        
        const [rows] = await bigquery.query({
            query: query,
            location: 'US'
        });
        
        res.status(200).json(rows);
        
    } catch (error) {
        console.error('Error querying BigQuery:', error);
        res.status(500).json({
            error: 'Failed to load statistics',
            message: error.message
        });
    }
}

// Export for Code Puppy Pages
module.exports = {
    getDistributionLists,
    searchDistributionLists,
    getDistributionListsByCategory,
    getDistributionListStats
};

File: Store Support/General Setup/Distribution_Lists/02_SCRIPTS_AND_TOOLS/api_distribution_lists.py
"""
Code Puppy Pages API Endpoint (Python/Flask)
Queries BigQuery for distribution lists

Deploy this to Code Puppy Pages as: /api/distribution-lists
"""

from flask import Flask, jsonify, request
from google.cloud import bigquery
import logging

# Initialize Flask app
app = Flask(__name__)

# Initialize BigQuery client
bigquery_client = bigquery.Client(project='wmt-assetprotection-prod')

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


@app.route('/api/distribution-lists', methods=['GET'])
def get_distribution_lists():
    """
    GET /api/distribution-lists
    Returns all distribution lists from BigQuery
    """
    try:
        query = """
            SELECT 
                email,
                name,
                display_name,
                description,
                member_count,
                category
            FROM `wmt-assetprotection-prod.Store_Support_Dev.dl_catalog`
            ORDER BY name
        """
        
        logger.info('Executing BigQuery query...')
        query_job = bigquery_client.query(query)
        results = query_job.result()
        
        # Convert results to list of dictionaries
        rows = []
        for row in results:
            rows.append({
                'email': row.email,
                'name': row.name,
                'display_name': row.display_name,
                'description': row.description,
                'member_count': row.member_count,
                'category': row.category
            })
        
        logger.info(f'Retrieved {len(rows)} distribution lists')
        
        return jsonify(rows), 200
        
    except Exception as error:
        logger.error(f'Error querying BigQuery: {str(error)}')
        return jsonify({
            'error': 'Failed to load distribution lists',
            'message': str(error)
        }), 500


@app.route('/api/distribution-lists/search', methods=['GET'])
def search_distribution_lists():
    """
    GET /api/distribution-lists/search?q=searchTerm
    Search distribution lists by keyword
    """
    try:
        search_term = request.args.get('q', '')
        
        query = """
            SELECT 
                email,
                name,
                display_name,
                description,
                member_count,
                category
            FROM `wmt-assetprotection-prod.Store_Support_Dev.dl_catalog`
            WHERE LOWER(email) LIKE @search_term
               OR LOWER(name) LIKE @search_term
               OR LOWER(description) LIKE @search_term
            ORDER BY name
            LIMIT 100
        """
        
        job_config = bigquery.QueryJobConfig(
            query_parameters=[
                bigquery.ScalarQueryParameter("search_term", "STRING", f"%{search_term.lower()}%")
            ]
        )
        
        query_job = bigquery_client.query(query, job_config=job_config)
        results = query_job.result()
        
        rows = []
        for row in results:
            rows.append({
                'email': row.email,
                'name': row.name,
                'display_name': row.display_name,
                'description': row.description,
                'member_count': row.member_count,
                'category': row.category
            })
        
        return jsonify(rows), 200
        
    except Exception as error:
        logger.error(f'Error searching BigQuery: {str(error)}')
        return jsonify({
            'error': 'Failed to search distribution lists',
            'message': str(error)
        }), 500


@app.route('/api/distribution-lists/stats', methods=['GET'])
def get_distribution_list_stats():
    """
    GET /api/distribution-lists/stats
    Get statistics about distribution lists
    """
    try:
        query = """
            SELECT 
                category,
                COUNT(*) as count,
                SUM(member_count) as total_members,
                AVG(member_count) as avg_members
            FROM `wmt-assetprotection-prod.Store_Support_Dev.dl_catalog`
            GROUP BY category
            ORDER BY count DESC
        """
        
        query_job = bigquery_client.query(query)
        results = query_job.result()
        
        rows = []
        for row in results:
            rows.append({
                'category': row.category,
                'count': row['count'],
                'total_members': row.total_members,
                'avg_members': float(row.avg_members) if row.avg_members else 0
            })
        
        return jsonify(rows), 200
        
    except Exception as error:
        logger.error(f'Error querying BigQuery: {str(error)}')
        return jsonify({
            'error': 'Failed to load statistics',
            'message': str(error)
        }), 500


# Health check endpoint
@app.route('/health', methods=['GET'])
def health_check():
    """Health check endpoint"""
    return jsonify({'status': 'healthy', 'service': 'distribution-lists-api'}), 200


if __name__ == '__main__':
    app.run(host='0.0.0.0', port=8080, debug=False)
